src/libcore/iter/iterator.rs

```diff
@@ -1368,6 +1368,7 @@ pub trait Iterator {
     /// [`Result`]: ../../std/result/enum.Result.html
     #[inline]
     #[stable(feature = "rust1", since = "1.0.0")]
+    #[must_use = "if you really need to exhaust the iterator, consider `.for_each(drop)` instead"]
```

Comment:
I think this is going to be annoying for the "collect to unit" use-case. I don't know how much we want to support that, but I thought it worth pointing out.

Reply:
Unit-returning functions never trigger the warning, even if they are generic in output type.  Also, there's a use case for that FromIterator implementation besides just making `collect::<Result<()>>` work?